src/mds/MDSRank.cc

```diff
@@ -3098,7 +3148,69 @@ void MDSRankDispatcher::handle_asok_command(
   } else if (command == "quiesce db") {
     command_quiesce_db(cmdmap, on_finish);
     return;
-  } else {
+  } 
+#ifdef WITH_CEPHFS_NOTIFICATION
+  else if (command == "add_topic") {
+    std::string topic_name, broker, username;
+    std::string password;
+    bool use_ssl;
+    std::optional<std::string> ca_location, mechanism;
+    cmd_getval(cmdmap, "topic_name", topic_name);
+    cmd_getval(cmdmap, "broker", broker);
+    if (!cmd_getval(cmdmap, "use_ssl", use_ssl)) {
+      use_ssl = false;
+    }
+    cmd_getval(cmdmap, "username", username);
+    cmd_getval(cmdmap, "password", password);
+    std::string ca, mch;
+    if (cmd_getval(cmdmap, "ca_location", ca)) {
+      ca_location = ca;
+    }
+    if (cmd_getval(cmdmap, "mechanism", mch)) {
+      mechanism = mch;
+    }
+    auto m = make_message<MNotificationInfoKafkaTopic>(
+        topic_name, broker, use_ssl, username, password, ca_location,
+        mechanism);
+    r = notification_manager->add_kafka_topic(topic_name, broker, use_ssl,
+                                              username, password, ca_location,
+                                              mechanism, true);
+    if (r == 0) {
+      send_notification_info_to_peers(m);
+    }
+  }
+  else if (command == "remove_topic") {
+    std::string topic_name;
+    cmd_getval(cmdmap, "topic_name", topic_name);
+    auto m = make_message<MNotificationInfoKafkaTopic>(topic_name, true);
+    r = notification_manager->remove_kafka_topic(topic_name, true);
+    if (r == 0) {
+      send_notification_info_to_peers(m);
+    }
+  }
+  else if (command == "add_udp_endpoint") {
+    std::string ip, name;
+    int64_t port;
+    cmd_getval(cmdmap, "entity", name);
+    cmd_getval(cmdmap, "ip", ip);
+    cmd_getval(cmdmap, "port", port);
+    auto m = make_message<MNotificationInfoUDPEndpoint>(name, ip, port);
```

Comment:
May be build and send out such notification messages inside add_\*_endpoint? Hence avoiding any reference to MNotificationInfo* at MDSRank.cc?